examples/qtmegachatapi/megachatapplication.cpp

```diff
@@ -0,0 +1,387 @@
+    "border-radius: 2px;"
+    "background-color: qlineargradient(spread:pad, x1:0, y1:0, x2:0, y2:1,"
+        "stop:0 rgba(100,100,100,255),"
+        "stop:1 rgba(160,160,160,255));");
+}
+
+
+
+
+AppDelegate appDelegate;
+const QEvent::Type GcmEvent::type = (QEvent::Type)QEvent::registerEventType();
+
+bool AppDelegate::event(QEvent* event)
+{
+    if (event->type() != GcmEvent::type)
+        return false;
+
+    megaProcessMessage(static_cast<GcmEvent*>(event)->ptr);    
+    return true;
+}
+
+void AppDelegate::onAppTerminate(MegaChatApplication *megaChatApp)
+{
+    static bool called = false;
+    if (called)
+        return;
+    called = true;   
+    //delete megaChatApp;
+}
+
+void AppDelegate::onEsidLogout()
+{/*
+    marshallCall([this]()
+    {
+        QObject::disconnect(qApp, SIGNAL(lastWindowClosed()), &appDelegate, SLOT(onAppTerminate()));
+
+        gClient.reset();
+        remove((gAppDir+"/sid").c_str());
+        delete mainWin;
+        QMessageBox::critical(nullptr, tr("Logout"), tr("Your session has been closed remotely"));
+        createWindowAndClient();
+
+        gClient->loginSdkAndInit(nullptr)
+        .then([]()
+        {
+            KR_LOG_DEBUG("New client initialized with new session");
+            saveSid(gSdk->dumpSession());
+            QObject::connect(qApp, SIGNAL(lastWindowClosed()), &appDelegate, SLOT(onAppTerminate()));
+            gClient->connect(Presence::kInvalid);
+        })
+        .fail([](const promise::Error& err)
+        {
+            KR_LOG_ERROR("Error re-creating or logging in chat client after ESID: ", err.what());
+        });
+    }, NULL);*/
+}
+
+extern "C" void AppDelegate::myMegaPostMessageToGui(void* msg, void* appCtx, AppDelegate *appDelegate)
+{
+    QEvent* event = new GcmEvent(msg);
+    QApplication::postEvent(appDelegate, event);
+}
+
+
+MegaChatApplication::MegaChatApplication(std::string &dir)
+{
+         //Create Dir
+         //appDir = karere::createAppDir();
+
+         appDir=dir;
+
+         //Create Main Window
+         mainWin = new MainWindow();
+
+         //Create MegaAPI
+         megaApi=new ::mega::MegaApi("karere-native", appDir.c_str(), "Karere Native");
+         megaApi->setLogLevel(MegaApi::LOG_LEVEL_DEBUG);
+         megaApi->addListener(this);
+
+         //Create MegaChatAPI
+         megaChatApi=new ::megachat::MegaChatApi(megaApi);
+         megaChatApi->setLogLevel(MegaChatApi::LOG_LEVEL_DEBUG);
+         megaChatApi->addChatRequestListener(this);
+         megaChatApi->addChatListener(this);
+
+         #ifndef KARERE_DISABLE_WEBRTC
+            megaChatApi->addChatCallListener(mainWin);
+         #endif
+
+         mainWin->setMegaChatApi(megaChatApi);
+
+         //Create Login Dialog
+         loginDlg = new LoginDialog(nullptr);
+
+         //Create websocket
+         websocketsIO = new LibwsIO();
+}
+
+MegaChatApplication::~MegaChatApplication()
+{
+    megaApi->logout(nullptr);
+    megaChatApi->logout(nullptr);
```

Comment:
There's a parameter by default already set to `NULL`. Actually, I think you better create a `logout()` facility at `MegaChatApplication`, so it can be called from here and other places in the app, in case of required. That facility can do extra stuff related to the login :)

Reply:
Done